### PR TITLE
[FLINK-20887][table-planner] Disable project merge during sql2rel phase by default to avoid incorrectly project merge

### PR DESCRIPTION
**Pull request by lincoln-lil**
## What is the purpose of the change
FLINK-30841 fixes incorrect project merge in the optimization phase (prevents projects containing non-deterministic calls from being merged with 'wrong results' that are not expected by users, also FLINK-15366 & FLINK-30006 had taken some efforts), but didn't fix the problem completely, as in the case described in this issue, we need to turn off the relevant optimizations in the sql2rel phase(add internal config and turn off by default) to fix it completely.

some more detailed explanation of the relate case, if we keep project merge in sql2rel phase, we won't have any chance to fix it because the projects are been merge already, see the original relnode tree after sql2rel:
```
== Abstract Syntax Tree ==
LogicalProject(exprs=[[-(+(RAND(), 7), +(RAND(), 5))]])
+- LogicalTableScan(table=[[default_catalog, default_database, SmallTable3]])
```

Note: this is not a new feature(the newly added config option is not recommended to use, only used as a rollback configuration when users need to restore the behavior of an older version)

## Brief change log
- 1. pre-step: turn off the project merge optimization in the sql2rel phase by default
- 2. fix incorrectly calc merge (but need to change related rules)
- 3. fix the fail case due to incorrect structured type's nullability (more details can see FLINK-31830)
- 4. fix the incorrectly aggregate project merge for LogicalWindowAggregate

## Verifying this change
existing and newly added cases 

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with @Public(Evolving): (yes)
  - The serializers: (no )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation
  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRuleTest.java [hunk 1]
```diff
@@ -352,8 +358,25 @@ public void testProjectionIncludingOnlyMetadata() {
...
+                FlinkHepRuleSetProgramBuilder.<BatchOptimizeContext>newBuilder()
+                        .setHepRulesExecutionType(HEP_RULES_EXECUTION_TYPE.RULE_SEQUENCE())
+                        .setHepMatchOrder(HepMatchOrder.BOTTOM_UP)
+                        .add(
+                                RuleSets.ofList(
+                                        CoreRules.PROJECT_MERGE,
+                                        PushProjectIntoTableSourceScanRule.INSTANCE))
+                        .build());
+        util().replaceBatchProgram(programs);
+    }
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRuleTest.java:374]: Why do we need to add this method in the pre-step commit？ Maybe it need to move to the Step2.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectMergeRule.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.apache.flink.table.planner.plan.utils.FlinkRexUtil;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.rules.ProjectMergeRule;
+
+/**
+ * Extends calcite's FilterCalcMergeRule for streaming scenario, modification: does not merge the
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectMergeRule.java:29]: Extends calcite's ProjectMergeRule

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectCalcMergeRule.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        project.getProjects().forEach(p -> p.accept(inputRefVisitor));
+        boolean existNonDeterministicRef =
+                Arrays.stream(inputRefVisitor.getFields())
+                        .anyMatch(i -> !RexUtil.isDeterministic(expandProjects.get(i)));
+
+        if (!existNonDeterministicRef) {
+            super.onMatch(call);
+        }
+    }
+}
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectCalcMergeRule.java:66]: is it more reasonable to re-implement matches() method  here, which to be consistent with ProjectMergeRule and CalcMergeRule.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala [hunk 4]
```diff
@@ -650,6 +654,75 @@ object FlinkRexUtil {
...
+  private class InputRefCounter(deep: Boolean, val refCounts: Array[Int])
+    extends RexVisitorImpl[Void](deep: Boolean) {
+    override def visitInputRef(inputRef: RexInputRef): Void = {
+      val index = inputRef.getIndex
+      refCounts(index) += 1
+      null
+    }
+  }
+
+  private def mergeable(
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:685]: Add some commets?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala [hunk 5]
```diff
@@ -650,6 +654,75 @@ object FlinkRexUtil {
...
+      topProgram.getProjectList.map(topProgram.expandLocalRef)
+    }.toList
+
+    mergeable(
+      topInputRefCounter,
+      topInputRefs.toList,
+      bottomProgram.getProjectList
+        .map(bottomProgram.expandLocalRef)
+        .toList)
+  }
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:725]: Nits: To achieve the goal of scala-free, would it be better to put these utils methods in a Java class?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala [hunk 6]
```diff
@@ -650,6 +654,75 @@ object FlinkRexUtil {
...
+      Array.fill(topProject.getInput.getRowType.getFieldCount)(0)
+
+    mergeable(topInputRefCounter, topProject.getProjects, bottomProject.getProjects)
+  }
+
+  /**
+   * An InputRefCounter that count every inputRef's reference count number, every reference will be
+   * counted, e.g., '$0 + 1' & '$0 + 2' will count 2 instead of 1.
+   * @param deep
+   * @param refCounts
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:674]: Missing Parameter Description

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectCalcMergeRule.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+        List<RexNode> expandProjects =
+                calc.getProgram().getProjectList().stream()
+                        .map(p -> calc.getProgram().expandLocalRef(p))
+                        .collect(Collectors.toList());
+        InputRefVisitor inputRefVisitor = new InputRefVisitor();
+        project.getProjects().forEach(p -> p.accept(inputRefVisitor));
+        boolean existNonDeterministicRef =
+                Arrays.stream(inputRefVisitor.getFields())
+                        .anyMatch(i -> !RexUtil.isDeterministic(expandProjects.get(i)));
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectCalcMergeRule.java:60]: Can resue FlinkRexUtil.Mergeable()?

**Review by swuferhong — COMMENTED** [review]
Thanks for the PR. It generally LGTM. Just leave some minor comments.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectMergeRule.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final Project topProject = call.rel(0);
+        final Project bottomProject = call.rel(1);
+        if (FlinkRexUtil.isMergeable(topProject, bottomProject)) {
+            super.onMatch(call);
+        }
+    }
```
Comment (swuferhong) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectMergeRule.java:47]: Ditto: is it more reasonable to re-implement matches() method here ?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectCalcMergeRule.java (hunk 7 above)
Comment (lincoln-lil) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkProjectCalcMergeRule.java]: From a maintainability point of view, it is indeed better to use a unified reusable logic here

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala (hunk 5 above)
Comment (lincoln-lil) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala]: Ok, this encourage me creating a new class `FlinkRelUtil` which is more appropriately than the `FlinkRexUtil` here

**Review by lincoln-lil — COMMENTED** [review]
@swuferhong thank you for reviewing this! I've addressed your comments and updated the pr

**Review by swuferhong — APPROVED** [review]
Thanks, @lincoln-lil . This PR LGTM now.

**Comment by lincoln-lil** [comment]
The squashed commits after rebased master branch succeed in my private pipeline: https://dev.azure.com/lincoln86xy/flink/_build/results?buildId=537&view=results

**Pull request closed without merging**